### ISSUE TITLE
Implement bindings for streaming crypto_auth functions

### ISSUE THREAD

**Issue by tarcieri** [dnaq/sodiumoxide#87]
I am using sodiumoxide in [rust-macaroons](https://github.com/cryptosphere/rust-macaroons).

The Macaroons format, as it works today, depends on HMAC supporting variable-length keys via hashing the key with the specified hash function in the event it is not the desired length. Lack of support for this in sodiumoxide has lead to a few unsightly hacks:

* https://github.com/cryptosphere/rust-macaroons/blob/master/src/token.rs#L12
* https://github.com/cryptosphere/rust-macaroons/blob/master/src/token.rs#L74 (effectively `bytes::copy_memory`, but it shouldn't need a copy)

I think sodiumoxide should support variable-sized byte arrays for HMAC keys. That's how the HMAC function is defined. I don't necessarily agree with it being defined that way and think it's a little weird, but it's making it hard for me to write real-world code and that's why I request it be changed.

**Comment by dnaq** [comment]
Hi,

I agree that a general implementation of HMAC should follow the HMAC standard. However the `crypto_auth` API in sodium/NaCl isn't a general HMAC implementation, it is an API for doing cryptographic authentication of messages under a secret key. The fact that the implementations of these API's use HMAC under the hood is more of an implementation detail. For example the `crypto_onetimeauth_poly1305` primitive exposes the exact same API, but uses a completely different cryptographic primitive under the hood.

What I might be able to do though, that you could use as a workaround to reduce unnecessary copying, is to add methods with the following signature to all newtypes (`Key`s, `Nonce`s, et.al.) `fn as_$newtype(bs: &'a [u8]) -> Option<&'a $newtype>` and `fn as_slice(&'a $newtype) -> &'a [u8]` that can coerce a slice into a newtype and back without copying (if the length of the slice is the same as the length of the newtype).

**Comment by jedisct1** [comment]
The `crypto_auth_hmacsha*()` functions follow the HMAC specification and support any key size.

Ex:
```c
int crypto_auth_hmacsha512256_init(crypto_auth_hmacsha512256_state *state,
                                   const unsigned char *key,
                                   size_t keylen);
```

**Comment by dnaq** [comment]
@jedisct1 thanks, that is indeed correct, maybe I should implement bindings for those functions as well.

**Comment by dnaq** [comment]
I pushed my code and tests to the `streaming_hmac` branch, haven't had the time to audit myself yet so I haven't merged it into master, but please take a look and see if it solves your needs.

**Comment by tarcieri** [comment]
Will take a look. Thanks!

**Comment by tarcieri** [comment]
@dnaq I finally ended up using this in `rust-macaroons` and it looks good. Maybe the only thing that confused me was why `State::init` doesn't use a `Key`.

Anyway, can you release a new version with this feature? Thanks!

**Comment by dnaq** [comment]
The reason is that `State::init` can take variable length keys, since it's a generic HMAC. I'm considering moving the streaming functions to a separate module (e.g. `crypto::hmac`) to make that clearer.

**Comment by tarcieri** [comment]
Aah! Makes sense.

**Comment by dnaq** [comment]
Released as sodiumoxide-0.0.9.

**Comment by tarcieri** [comment]
Sweet, thank you!
